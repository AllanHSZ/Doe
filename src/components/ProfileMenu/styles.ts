import styled, { css, DefaultTheme } from 'styled-components'
import media from 'styled-media-query'

export const Nav = styled.nav`
  ${({ theme }) => css`
    display: flex;
    box-shadow: ${theme.boxShadow.medium};
    border-bottom: 0.1rem solid ${theme.colors.lightGray};
    height: fit-content;

    @media screen and (min-width: 960px) {
      flex-direction: column;
      border: 0;
      a:not(:last-child) {
        border-bottom: 0.1rem solid ${theme.colors.lightGray};
      }
    }
  `}
`

const linkModifiers = {
  default: (theme: DefaultTheme) => css`
    background: ${theme.colors.white};
    color: ${theme.colors.text};
  `,

  active: (theme: DefaultTheme) => css`
    background: ${theme.colors.primary};
    color: ${theme.colors.white};
  `
}

type LinkProps = {
  isActive?: boolean
}

export const Link = styled.a<LinkProps>`
  ${({ theme, isActive }) => css`
    flex: 1;
    justify-content: center;
    background: ${theme.colors.white};
    color: ${theme.colors.text};
    text-decoration: none;
    display: flex;
    align-items: center;
    padding: ${theme.spacings.xsmall} ${theme.spacings.small};
    transition: background, color, ${theme.transition.default};

    &:hover {
      background: ${theme.colors.primary};
      color: ${theme.colors.white};
    }
    > span {
      display: none;
      margin-left: ${theme.spacings.xsmall};
    }

    ${media.greaterThan('medium')`
      /* flex: unset; */
      justify-content: initial;

      > span {
        display: initial;
      }
    `}
    ${!isActive && linkModifiers.default(theme)};
    ${isActive && linkModifiers.active(theme)};
  `}
`
