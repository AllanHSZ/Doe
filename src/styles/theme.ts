export default {
  grid: {
    container: '110rem',
    gutter: '3.2rem'
  },
  border: {
    radius: '0.6rem'
  },
  font: {
    family:
      "Lato, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
    light: 300,
    normal: 400,
    bold: 700,
    sizes: {
      xsmall: '1.2rem',
      small: '1.4rem',
      medium: '1.6rem',
      large: '1.8rem',
      xlarge: '2.0rem',
      xxlarge: '2.8rem',
      huge: '4.0rem'
    }
  },
  colors: {
    primary: '#04D361',
    primaryDark: '#04AB4F',
    secondary: '#D30476',
    textEmphasis: '#212D22',
    text: '#394338',
    white: '#FAFAFA',
    lightGray: '#F2F3F5',
    gray: '#8F8F8F',
    darkGray: '#2E2F42',
    red: '#E33D3D'
    // background: ''
    // mainBg: '#06092B',
    // lightBg: '#F2F2F2',
    // black: '#030517',
  },
  spacings: {
    xxxsmall: '0.4rem',
    xxsmall: '0.8rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem'
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50
  },
  transition: {
    default: '0.3s ease-in-out',
    fast: '0.1s ease-in-out'
  }
} as const
