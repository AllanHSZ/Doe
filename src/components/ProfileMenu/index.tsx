import Link from 'next/link'
import { Archive, Heart, LogOut, MapPin, User } from 'react-feather'
import { signOut } from 'next-auth/client'
import { useRouter } from 'next/router'

import * as S from './styles'

export type ProfileMenuProps = {
  activeLink?: '/profile/me' | '/profile/address' | '/profile/donation' | string
}

const ProfileMenu = ({ activeLink }: ProfileMenuProps) => {
  const router = useRouter()

  return (
    <S.Nav>
      <Link href="/profile/me" passHref>
        <S.Link isActive={activeLink === '/profile/me'} title="Meus dados">
          <User size={24} />
          <span>Meus dados</span>
        </S.Link>
      </Link>

      <Link href="/profile/address" passHref>
        <S.Link
          isActive={activeLink === '/profile/address'}
          title="Meus endereços"
        >
          <MapPin size={24} />
          <span>Meus endereços</span>
        </S.Link>
      </Link>

      <Link href="/profile/donations" passHref>
        <S.Link
          isActive={activeLink === '/profile/donations'}
          title="Minhas doações"
        >
          <Heart size={24} />
          <span>Minhas doações</span>
        </S.Link>
      </Link>

      <Link href="/profile/received" passHref>
        <S.Link
          isActive={activeLink === '/profile/received'}
          title="Minhas reservas"
        >
          <Archive size={24} />
          <span>Minhas reservas</span>
        </S.Link>
      </Link>

      <S.Link
        role="button"
        onClick={async () => {
          const data = await signOut({ redirect: false, callbackUrl: '/' })
          router.push(data.url)
        }}
      >
        <LogOut size={24} aria-label="Sair" />
        <span>Sair</span>
      </S.Link>
    </S.Nav>
  )
}

export default ProfileMenu
