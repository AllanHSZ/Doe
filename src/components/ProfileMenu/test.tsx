import { screen } from '@testing-library/react'
import theme from 'styles/theme'
import { renderWithTheme } from 'utils/tests/helpers'

import ProfileMenu from '.'

// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/router'), 'useRouter')

useRouter.mockImplementation(() => ({
  query: {}
}))

describe('<ProfileMenu />', () => {
  it('should render the menu', () => {
    const { container } = renderWithTheme(<ProfileMenu />)

    expect(
      screen.getByRole('link', { name: /meus dados/i })
    ).toBeInTheDocument()

    expect(
      screen.getByRole('link', { name: /meus endereços/i })
    ).toBeInTheDocument()
    expect(
      screen.getByRole('link', { name: /minhas Doações/i })
    ).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /sair/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()

    expect(container.firstChild).toMatchSnapshot()
  })

  it(`should render the menu with an active link defined`, () => {
    renderWithTheme(<ProfileMenu activeLink="/profile/address" />)

    expect(screen.getByRole('link', { name: /meus endereços/i })).toHaveStyle({
      background: theme.colors.primary,
      color: theme.colors.white
    })
  })
})
