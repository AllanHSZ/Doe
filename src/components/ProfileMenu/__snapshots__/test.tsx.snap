// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileMenu /> should render the menu 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-shadow: 0px 0px 8px rgba(0,0,0,0.15);
  border-bottom: 0.1rem solid #F2F3F5;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #FAFAFA;
  color: #394338;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1.6rem 2.4rem;
  -webkit-transition: background,color,0.3s ease-in-out;
  transition: background,color,0.3s ease-in-out;
  background: #FAFAFA;
  color: #394338;
}

.c1:hover {
  background: #04D361;
  color: #FAFAFA;
}

.c1 > span {
  display: none;
  margin-left: 1.6rem;
}

@media screen and (min-width:960px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    border: 0;
  }

  .c0 a:not(:last-child) {
    border-bottom: 0.1rem solid #F2F3F5;
  }
}

@media (min-width:768px) {
  .c1 {
    -webkit-box-pack: initial;
    -webkit-justify-content: initial;
    -ms-flex-pack: initial;
    justify-content: initial;
  }

  .c1 > span {
    display: initial;
  }
}

<nav
  class="c0"
>
  <a
    class="c1"
    href="/profile/me"
    title="Meus dados"
  >
    <svg
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20 21v-2a4 4 0 0 0-4-4H8a4 4 0 0 0-4 4v2"
      />
      <circle
        cx="12"
        cy="7"
        r="4"
      />
    </svg>
    <span>
      Meus dados
    </span>
  </a>
  <a
    class="c1"
    href="/profile/address"
    title="Meus endereços"
  >
    <svg
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 10c0 7-9 13-9 13s-9-6-9-13a9 9 0 0 1 18 0z"
      />
      <circle
        cx="12"
        cy="10"
        r="3"
      />
    </svg>
    <span>
      Meus endereços
    </span>
  </a>
  <a
    class="c1"
    href="/profile/donations"
    title="Minhas Doações"
  >
    <svg
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20.84 4.61a5.5 5.5 0 0 0-7.78 0L12 5.67l-1.06-1.06a5.5 5.5 0 0 0-7.78 7.78l1.06 1.06L12 21.23l7.78-7.78 1.06-1.06a5.5 5.5 0 0 0 0-7.78z"
      />
    </svg>
    <span>
      Minhas Doações
    </span>
  </a>
  <a
    class="c1"
    role="button"
  >
    <svg
      aria-label="Sair"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
      />
      <polyline
        points="16 17 21 12 16 7"
      />
      <line
        x1="21"
        x2="9"
        y1="12"
        y2="12"
      />
    </svg>
    <span>
      Sair
    </span>
  </a>
</nav>
`

exports[`<ProfileMenu /> should render the menu 2`] = `
<nav
  class="sc-bdvvtL dckMoB"
>
  <a
    class="sc-gsDKAQ QdZjM"
    href="/profile/me"
    title="Meus dados"
  >
    <svg
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20 21v-2a4 4 0 0 0-4-4H8a4 4 0 0 0-4 4v2"
      />
      <circle
        cx="12"
        cy="7"
        r="4"
      />
    </svg>
    <span>
      Meus dados
    </span>
  </a>
  <a
    class="sc-gsDKAQ QdZjM"
    href="/profile/address"
    title="Meus endereços"
  >
    <svg
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 10c0 7-9 13-9 13s-9-6-9-13a9 9 0 0 1 18 0z"
      />
      <circle
        cx="12"
        cy="10"
        r="3"
      />
    </svg>
    <span>
      Meus endereços
    </span>
  </a>
  <a
    class="sc-gsDKAQ QdZjM"
    href="/profile/donations"
    title="Minhas Doações"
  >
    <svg
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20.84 4.61a5.5 5.5 0 0 0-7.78 0L12 5.67l-1.06-1.06a5.5 5.5 0 0 0-7.78 7.78l1.06 1.06L12 21.23l7.78-7.78 1.06-1.06a5.5 5.5 0 0 0 0-7.78z"
      />
    </svg>
    <span>
      Minhas Doações
    </span>
  </a>
  <a
    class="sc-gsDKAQ QdZjM"
    role="button"
  >
    <svg
      aria-label="Sair"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
      />
      <polyline
        points="16 17 21 12 16 7"
      />
      <line
        x1="21"
        x2="9"
        y1="12"
        y2="12"
      />
    </svg>
    <span>
      Sair
    </span>
  </a>
</nav>
`
